Reads a specific sheet from the Excel file based on XlsxReadOptions sheetIndex

emilianbold opened the pull request:
…sheetIndex

Thanks for contributing.

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](https://developercertificate.org)

## Description

Adds XlsxReadOptions.tableIndex to select which table to return by default in `read(XlsxReadOptions)`. Somewhat in style with the HtmlReadOptions flag.

I wanted this to be a `sheetIndex` but then noticed some sheets might not have tables so the index wouldn't match...

## Testing

Yes.

benmccann commented:
There's a typo in "file" in the commit message

benmccann commented:
There's a typo in "file" in the commit message. Also the commit message says `sheetIndex` but then the code uses `tableIndex`

benmccann commented on excel/src/main/java/tech/tablesaw/io/xlsx/XlsxReadOptions.java:
```diff
@@ -12,8 +12,12 @@
 
 public class XlsxReadOptions extends ReadOptions {
 
+  /** Which table should be read? Null means no specific index was set. First table has index 0. */
```
I think it sounds funny to put a question in the javadoc. How about "The table to read" instead of "Which table should be read?"

benmccann commented:
I also like the idea of `sheetIndex` instead of `tableIndex`. You could make that work by having `readMultiple` return a list containing `null` values when there's no table in a sheet or a `Map<Integer, Table>` with the sheet number

What do you think @hallvard? I'd like to get your opinion since you wrote this code initially

emilianbold commented:
You are right, `sheetIndex` is what I originally wanted too. I've updated the commit, fixed the message and added more tests.

Picked the idea of adding `null`s for empty sheets, but I do this in a protected method so the public `readMultiple` method remains unchanged for users.

benmccann commented on excel/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java:
```diff
@@ -60,14 +61,47 @@ public static void register(ReaderRegistry registry) {
+        }
+        return table;
+      } else {
+        throw new IndexOutOfBoundsException(
```
How about:

```
if (options.sheetIndex() < 0 || options.sheetIndex() >= tables.size()) {
  throw new IndexOutOfBoundsException(
    "Sheet index outside bounds. " + tables.size() + " sheets found.");
}
```

Then the rest of the code doesn't need to be nested in the `if/else` statement

benmccann commented on excel/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java:
```diff
@@ -60,14 +61,47 @@ public static void register(ReaderRegistry registry) {
+      }
+    }
+    // since no specific sheetIndex asked, return first table
+    Optional<Table> table = tables.stream().filter(t -> t != null).findFirst();
```
Could do:

```
return tables.stream().filter(t -> t != null).findFirst()
    .orElseThrow(() -> new IllegalArgumentException("No tables found."));
```

benmccann commented on excel/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java:
```diff
@@ -60,14 +61,47 @@ public static void register(ReaderRegistry registry) {
+  /**
+   * Read at most a table from every sheet.
+   *
+   * @param options
```
should either fill out or remove. probably the same with the `@throws` documentation

benmccann commented:
Great idea! I like it! Just a few minor comments

emilianbold commented:
OK, updated the PR.

emilianbold commented:
I think it's looking pretty good. Do I need to change something else?

benmccann commented on excel/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java:
```diff
@@ -60,14 +60,42 @@ public static void register(ReaderRegistry registry) {
   public Table read(XlsxReadOptions options) throws IOException {
-    List<Table> tables = readMultiple(options);
+    List<Table> tables = readMultiple(options, true);
     if (tables.isEmpty()) {
```
You probably don't need this check anymore since there are other checks that would cover it as well

emilianbold replied:
:-)

benmccann commented on excel/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java:
```diff
@@ -60,14 +60,39 @@ public static void register(ReaderRegistry registry) {
+
+      Table table = tables.get(index);
+      if (table == null) {
+        throw new IllegalArgumentException("No tables found.");
```
Let's change this message to:

    throw new IllegalArgumentException("No table found at index " + index);

emilianbold replied:
Maybe something like `"No table found for sheet index " + index`?

emilianbold replied:
Or `found at sheet index`?

benmccann replied:
sure. "at sheet index" sounds good to me

benmccann commented on excel/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java:
```diff
@@ -60,14 +60,39 @@ public static void register(ReaderRegistry registry) {
+      int index = options.sheetIndex();
+      if (index < 0 || index >= tables.size()) {
+        throw new IndexOutOfBoundsException(
+            "Sheet index outside bounds. " + tables.size() + " sheets found.");
```
sorry. one last thing I just noticed. How about also adding the index here:

    "Sheet index " + index + " outside bounds. " + tables.size() + " sheets found.");

emilianbold replied:
Done. If you want I can also add a test for this:

`assertEquals("Sheet index 5 outside bounds. 2 sheets found.", iobe.getMessage());`

benmccann commented:
This is great! Thanks for all the help!
